### PR TITLE
Add readme for Data Virtualization and rename groupId

### PR DESCRIPTION
**Pull request by jatgarg**
## Description

This is part of data virtualization effort which is documented here: [Design Doc](https://microsoft.sharepoint.com/:w:/t/FluidFramework/ESJQvukxffJNiTZzbNCWky4B891mpOpWD7BHhbOznXsMZg?e=gAUCrx)

1.) Add readme for Data Virtualization.
2.) Rename groupId in datastore creation apis.

### CODE REVIEW

File: packages/runtime/container-runtime/api-report/container-runtime.api.md [hunk 1]
```diff
@@ -121,11 +121,11 @@ export class ContainerRuntime extends TypedEventEmitter<IContainerRuntimeEvents
     // (undocumented)
     get connected(): boolean;
     // (undocumented)
-    createDataStore(pkg: string | string[], groupId?: string): Promise<IDataStore>;
+    createDataStore(pkg: string | string[], dataStoreGroupIdForSnapshotFetch?: string): Promise<IDataStore>;
     // @deprecated (undocumented)
     _createDataStoreWithProps(pkg: string | string[], props?: any, id?: string): Promise<IDataStore>;
     // (undocumented)
-    createDetachedDataStore(pkg: Readonly<string[]>, groupId?: string): IFluidDataStoreContextDetached;
+    createDetachedDataStore(pkg: Readonly<string[]>, dataStoreGroupIdForSnapshotFetch?: string): IFluidDataStoreContextDetached;
```
Comment (vladsud) [inline packages/runtime/container-runtime/api-report/container-runtime.api.md:128]: I do not think a rename like that would help. If anything, I think it becomes even more confusing.
I'd personally keep the old name, but ensure that every place that introduces this concept has a link to proper documentation.

If we really want to put more descriptive name, then maybe loadingGroupId is a better term.
I'd definitely remove "dataStore" from the name as it's not adding any value (API is already on data store, and loading flow does not care where these IDs should up - on data stores or DDSs or somewhere else).

File: packages/runtime/runtime-definitions/src/dataStoreContext.ts [hunk 2]
```diff
@@ -195,21 +195,32 @@ export interface IContainerRuntimeBase extends IEventProvider<IContainerRuntimeB
 	 * already attached DDS (or non-attached DDS that will eventually get attached to storage) will result in this
 	 * store being attached to storage.
 	 * @param pkg - Package name of the data store factory
-	 * @param groupId - group to which this data stores belongs to. This is also known at service side and can be used to
-	 * fetch snapshot contents like snapshot tree, blobs using this id from the storage.
-	 */
-	createDataStore(pkg: string | string[], groupId?: string): Promise<IDataStore>;
+	 * @param dataStoreGroupIdForSnapshotFetch - This represents the group of the datastore within a container or its snapshot.
+	 * When not specified the datastore will belong to a `default` group. When a container is loaded initially, only datastores
```
Comment (vladsud) [inline packages/runtime/runtime-definitions/src/dataStoreContext.ts:199]: I'd suggest not to put any text here at all, and only have a reference to readme.md
Alternatively, have such block not repeated - have in one place, and just reuse documentation (use @inheritDoc tag)

File: packages/runtime/container-runtime/README.md [hunk 3]
```diff
@@ -51,6 +51,54 @@ issue](https://github.com/microsoft/FluidFramework/wiki/Submitting-Bugs-and-Feat
 
 Thank you!
 
+## Data Virtualization For DataStores
+
+This is the ability of container to delay load datastores at any point in time by asking for the datastore snapshot at
+anytime, not caring about from where it is coming and then using that to load that datastore and return it to the app.
```
Comment (vladsud) [inline packages/runtime/container-runtime/README.md:57]: This paragraph is hard to parse.
I'd say something like that
"It's a capability to exclude some content from initial snapshot (used when loading container) and thus improve boot performance of a container. Excluded content could be loaded at a later time when it's required."

File: packages/runtime/container-runtime/README.md [hunk 4]
```diff
@@ -51,6 +51,54 @@ issue](https://github.com/microsoft/FluidFramework/wiki/Submitting-Bugs-and-Feat
...
+limitations in interface and lack of appropriate semantics at driver API layer).
+
+### Improvement with Data Virtualization
+
+With this, we will provide a capability to:
+
+- Exclude some sub-trees from snapshot payload, thus allowing faster transfer times / boot times and smaller initial
+memory footprint.
+- Ability to safely (in current system, i.e., without introducing any new breaking behaviors) to delay-load data
+stores later.
```
Comment (vladsud) [inline packages/runtime/container-runtime/README.md:78]: I'd not mention safety - it's assumed implicitly that system works correctly and without bugs

File: packages/runtime/container-runtime/README.md [hunk 5]
```diff
@@ -51,6 +51,54 @@ issue](https://github.com/microsoft/FluidFramework/wiki/Submitting-Bugs-and-Feat
...
+
+Container Runtime Apis like IContainerRuntimeBase.createDataStore and IContainerRuntimeBase.createDetachedDataStore
+provides an argument `dataStoreGroupIdForSnapshotFetch` which allows apps to put a datastore in group at time of
```
Comment (vladsud) [inline packages/runtime/container-runtime/README.md:81]: Maybe "put" -> mark? I.e. we are marking data stores with a groupID, not putting them anywhere

Comment (vladsud) [reply packages/runtime/container-runtime/README.md:81]: I'd probably start that every data store is assigned a groupID. Not providing groupID (on API) means that default ID is used.

File: packages/runtime/container-runtime/README.md [hunk 6]
```diff
@@ -51,6 +51,54 @@ issue](https://github.com/microsoft/FluidFramework/wiki/Submitting-Bugs-and-Feat
...
+creation currently. This represents the group of the datastore within a container or its snapshot. When not specified
+the datastore will belong to a `default` group. When a container is loaded initially, only datastores which belongs to
+`default` group are fetched from service and can be loaded on demand when requested by user. This decreases the amount
+of data which needs to be fetched during load and hence provides faster boot times for the container. Snapshot for all
+datastores within a non-default datastore groupId will be fetched from service when any of the datastores within a
+group is requested by application, then snapshot for that particular group will be fetched using a network call at that
+time. So, 1 network will be required to fetch snapshot for a group when a datastore from a group is requested by an
```
Comment (vladsud) [inline packages/runtime/container-runtime/README.md:88]: 1 network -> "one network call"

File: packages/runtime/container-runtime/README.md [hunk 7]
```diff
@@ -51,6 +51,54 @@ issue](https://github.com/microsoft/FluidFramework/wiki/Submitting-Bugs-and-Feat
...
+application. This allows applications to put a datastore in a group, when it thinks that, that particular datastore
+will not be requested on boot and will be loaded later on like on some user action like a button click. This will
```
Comment (vladsud) [inline packages/runtime/container-runtime/README.md:90]: I'd refrain from using words "allows" and "when it thinks".
I'd use something like "when datastore is assigned to a group, content of such data store will not be loaded with initial load of container. It will be loaded only when any datastore with such groupID is realized.

Comment (vladsud) [reply packages/runtime/container-runtime/README.md:90]: same below - let's remove all mentions to "thinks" :)

File: packages/runtime/container-runtime/api-report/container-runtime.api.md (hunk 1 above)
Comment (jatgarg) [reply packages/runtime/container-runtime/api-report/container-runtime.api.md:128]: Renamed it to "loadingGroupId" on the api.

File: packages/runtime/runtime-definitions/src/dataStoreContext.ts (hunk 2 above)
Comment (jatgarg) [reply packages/runtime/runtime-definitions/src/dataStoreContext.ts:199]: I would like to keep doc on these apis as these are the base apis and everyone does does not go through a Readme. 
I will link to this place when using these somewhere else.

File: packages/runtime/container-runtime/README.md [hunk 8]
```diff
@@ -51,6 +51,54 @@ issue](https://github.com/microsoft/FluidFramework/wiki/Submitting-Bugs-and-Feat
...
+
+Fluid (through its APIs) does expose some levels of data virtualization. But under covers, there is almost none. The
+whole content of the file is downloaded in one go. Due to limitation of data virtualization, FF holds ALL blobs in
+snapshot as those might be required in the future. Any delayed loading (through FF APIs) results in loading state of
+DDSs at a sequence number of snapshot we booted from, and application of all the ops up till current sequence number.
+While application may choose not to load some data stores immediately on boot (and realize some saving in time and
+memory by not allocating appropriate app state for such datastores), FF still pays the costs for such content.
+It also continues to pay the cost for all such content indefinitely, even if those datastores were loaded (this could
+theoretically be optimized, but now there is no good way for driver to learn what blobs are safe to discard due to
+limitations in interface and lack of appropriate semantics at driver API layer).
```
[the remark this answers is not in the document]
Comment (jatgarg) [reply packages/runtime/container-runtime/README.md:69]: I could rename the heading and some text and put it in section like:
"Motivation for Data Virtualization" and mention what used to happen before Dv. It would be good for the reader to know about it.

[the remark this answers is not in the document]
Comment (jatgarg) [reply packages/runtime/container-runtime/README.md:69]: Maybe later when we are already in the DV world for some time, then we can remove it later on as I would think we would keep updating this readme as we implement more features and capabilities based on DV.

File: packages/runtime/container-runtime/README.md (hunk 4 above)
Comment (jatgarg) [reply packages/runtime/container-runtime/README.md:78]: done

File: packages/runtime/container-runtime/README.md (hunk 5 above)
Comment (jatgarg) [reply packages/runtime/container-runtime/README.md:81]: done

File: packages/runtime/container-runtime/README.md (hunk 6 above)
Comment (jatgarg) [reply packages/runtime/container-runtime/README.md:88]: done

File: packages/runtime/container-runtime/README.md (hunk 7 above)
Comment (jatgarg) [reply packages/runtime/container-runtime/README.md:90]: done

File: packages/runtime/container-runtime/README.md [hunk 9]
```diff
@@ -51,6 +51,54 @@ issue](https://github.com/microsoft/FluidFramework/wiki/Submitting-Bugs-and-Feat
...
+- Ability to safely (in current system, i.e., without introducing any new breaking behaviors) to delay-load data
+stores later.
+
+Container Runtime Apis like IContainerRuntimeBase.createDataStore and IContainerRuntimeBase.createDetachedDataStore
+provides an argument `dataStoreGroupIdForSnapshotFetch` which allows apps to put a datastore in group at time of
+creation currently. This represents the group of the datastore within a container or its snapshot. When not specified
+the datastore will belong to a `default` group. When a container is loaded initially, only datastores which belongs to
+`default` group are fetched from service and can be loaded on demand when requested by user. This decreases the amount
+of data which needs to be fetched during load and hence provides faster boot times for the container. Snapshot for all
+datastores within a non-default datastore groupId will be fetched from service when any of the datastores within a
```
[the remark this answers is not in the document]
Comment (jatgarg) [reply packages/runtime/container-runtime/README.md:86]: Made appropriate changes.

File: packages/runtime/container-runtime/README.md [hunk 10]
```diff
@@ -51,6 +51,52 @@ issue](https://github.com/microsoft/FluidFramework/wiki/Submitting-Bugs-and-Feat
...
 
+## Data Virtualization For DataStores (Work in Progress)
+
+It's a capability to exclude some content from initial snapshot (used when loading container) and thus improve boot
+performance of a container. Excluded content could be loaded at a later time when it's required.
+
+### Motivation for Data Virtualization
+
+This section talks about how the system used to work before Data virtualization. Fluid (through its APIs) does expose
+some levels of data virtualization. But under covers, there is almost none. The whole content of the file is downloaded
```
Comment (rajatch-ff) [inline packages/runtime/container-runtime/README.md:62]: What's the significance of this statement? If Fluid doesn't have any useful DV, what does "some level of DV" mean? If we don't have a good answer, let's just remove this.

File: packages/runtime/container-runtime/README.md [hunk 11]
```diff
@@ -51,6 +51,52 @@ issue](https://github.com/microsoft/FluidFramework/wiki/Submitting-Bugs-and-Feat
...
+in one go. Due to limitation of data virtualization, FF holds ALL blobs in snapshot as those might be required in the
+future. Any delayed loading (through FF APIs) results in loading state of DDSs at a sequence number of snapshot we
+booted from, and application of all the ops up till current sequence number. While application may choose not to load
```
Comment (rajatch-ff) [inline packages/runtime/container-runtime/README.md:65]: >and application of all the ops up till current sequence number. [](http://example.com/codeflow?start=13&length=63)

this sentence doesn't read/fit well with what precedes it.

File: packages/runtime/container-runtime/README.md [hunk 12]
```diff
@@ -51,6 +51,52 @@ issue](https://github.com/microsoft/FluidFramework/wiki/Submitting-Bugs-and-Feat
...
+## Data Virtualization For DataStores (Work in Progress)
+
+It's a capability to exclude some content from initial snapshot (used when loading container) and thus improve boot
+performance of a container. Excluded content could be loaded at a later time when it's required.
+
+### Motivation for Data Virtualization
+
+This section talks about how the system used to work before Data virtualization. Fluid (through its APIs) does expose
+some levels of data virtualization. But under covers, there is almost none. The whole content of the file is downloaded
+in one go. Due to limitation of data virtualization, FF holds ALL blobs in snapshot as those might be required in the
```
Comment (rajatch-ff) [inline packages/runtime/container-runtime/README.md:63]: >blobs [](http://example.com/codeflow?start=66&length=5)

blobs and datastores?

File: packages/runtime/container-runtime/README.md (hunk 11 above)
Comment (jatgarg) [reply packages/runtime/container-runtime/README.md:65]: Didn't get it? What part does not fit well?

File: packages/runtime/container-runtime/README.md (hunk 12 above)
Comment (jatgarg) [reply packages/runtime/container-runtime/README.md:63]: Datastore Context object is still there but we don't want to talk about that here in readme as that is internal concept.
Talking about blobs to user is fine as that is the main content of the snapshot which takes space.

On that note, I am reconsidering whether we want to tell what happened before as that will become obsolete once we have DV as Vlad mentioned. It is also making it a bit verbose.

File: packages/runtime/container-runtime/README.md [hunk 13]
```diff
@@ -51,6 +51,52 @@ issue](https://github.com/microsoft/FluidFramework/wiki/Submitting-Bugs-and-Feat
...
+belongs to `default` group are fetched from service and can be loaded on demand when requested by user. This decreases
+the amount of data which needs to be fetched during load and hence provides faster boot times for the container.
+Attempting to load any datastore within a non-default group results in fetching all content/datastores marked with same
+groupId and thus helps faster realization/loading of other datastores within that group faster.
+So, one network will be required to fetch content for a group when a datastore from a group is requested by an
+application.
+So to summarize, when datastore is assigned to a group, content of such data store will not be loaded with initial load
+of container. It will be loaded only when any datastore with such groupID is realized.
+This will improve the boot perf. Data virtualization or providing the `loadingGroupId` will however decrease the
+performance of loading of those datastores while improving the overall performance of container load. However,
```
Comment (rajatch-ff) [inline packages/runtime/container-runtime/README.md:93]: what does this mean?

File: packages/runtime/container-runtime/README.md [hunk 14]
```diff
@@ -51,6 +51,52 @@ issue](https://github.com/microsoft/FluidFramework/wiki/Submitting-Bugs-and-Feat
...
+With this, we will provide a capability to:
+
+- Exclude some sub-trees from snapshot payload, thus allowing faster transfer times / boot times and smaller initial
+memory footprint.
+- Ability to delay-load data stores later.
+
+Container Runtime Apis like IContainerRuntimeBase.createDataStore and IContainerRuntimeBase.createDetachedDataStore
+provides an argument `loadingGroupId` which allows apps to mark a datastore at time of creation currently. Every data
+store is assigned a groupID. Not providing groupID (on API) means that default ID is used. This groupId represents the
+group of the datastore within a container or its snapshot. When a container is loaded initially, only datastores which
```
[the remark this answers is not in the document]
Comment (jatgarg) [reply packages/runtime/container-runtime/README.md:83]: Yes, the app should only provide groupid to a datastore if it can get fairly big in size which we mentioned here in the readme. App should not provide groupid to just any datastore. It is ok if all datastores are in 1 group if the content in them is not supposed to be humongous like a subpage. This is what happens today as in all of the are in same group.

File: packages/runtime/container-runtime/api-report/container-runtime.api.md (hunk 1 above)
Comment (rajatch-ff) [reply packages/runtime/container-runtime/api-report/container-runtime.api.md]: What's the reason for not using 'groupId' alone?

File: packages/runtime/container-runtime/README.md (hunk 13 above)
Comment (jatgarg) [reply packages/runtime/container-runtime/README.md:93]: It means that DV will improve the boot perf of initial load of container but decrease the performance of loading of datastores with groupid as it will now involve making a network call while before no network call was required as the snapshot already existed in memory.

Comment (jatgarg) [reply packages/runtime/container-runtime/README.md:93]: I could try to make it clearer here in readme.

File: packages/runtime/runtime-definitions/src/dataStoreContext.ts (hunk 2 above)
Comment (rajatch-ff) [reply packages/runtime/runtime-definitions/src/dataStoreContext.ts:199]: I agree with Vlad on this. 'create' API annotation doesn't need to speak to how datastores are going to be 'fetched' using GroupId. That's more of a detail for the fetch logic not create logic. Another challenge with this annotation would be to always keep it up to date. When your readMe changes to reflect a new way of fetching, there's a high chance, the future engineer (or you) will forget to update this copy of comments.

**Review by rajatch5 — APPROVED** [review]
:shipit:

File: packages/runtime/container-runtime/README.md (hunk 12 above)
Comment (rajatch-ff) [reply packages/runtime/container-runtime/README.md:63]: Okay Thanks.
For the 2nd half, you can say "Without DV" this is what happens and "With DV" this is how it will happen.

File: packages/runtime/container-runtime/src/dataStores.ts [hunk 15]
```diff
@@ -438,7 +443,7 @@ export class DataStores implements IDisposable {
 			snapshotTree: undefined,
 			isRootDataStore: false,
 			createProps: props,
-			groupId,
+			groupId: loadingGroupId,
```
Comment (ChumpChief) [inline packages/runtime/container-runtime/src/dataStores.ts:446]: Also change name in IFluidDataStoreContextProps for consistent naming throughout.

**Review by ChumpChief — CHANGES_REQUESTED** [review]

File: packages/runtime/container-runtime/src/dataStores.ts (hunk 15 above)
Comment (ChumpChief) [reply packages/runtime/container-runtime/src/dataStores.ts:446]: And anywhere else (e.g. ISnapshotTree, etc.)

Comment (jatgarg) [reply packages/runtime/container-runtime/src/dataStores.ts:446]: Internally we would like to keep groupid, for api we can expose loadinggroupid if that makes it more understandable to the user. We should strive for smaller names in summaries and snapshots as they take space too. So lets keep groupid to be user internally and server contracts.

Comment (ChumpChief) [reply packages/runtime/container-runtime/src/dataStores.ts:446]: We do this in other places for other APIs and it's just confusing - if you are concerned about snapshot size, then you should deviate only at the snapshot layer and include detailed tsdoc that explains the name difference and the equivalence.  I disagree with this choice but wouldn't block it.

But IFluidDataStoreContextProps has no bearing on the snapshot and should be changed to match regardless.

Comment (jatgarg) [reply packages/runtime/container-runtime/src/dataStores.ts:446]: Yes, I will change here and in container runtime.

Comment (ChumpChief) [reply packages/runtime/container-runtime/src/dataStores.ts:446]: Also fwiw - even for the snapshot probably prefer something like lgid (loading group ID) for better parallel with the real name, more concise, less likely name collision with other "groupID" in the repo.

Comment (jatgarg) [reply packages/runtime/container-runtime/src/dataStores.ts]: Made changes in the container runtime to use loadingGroupId internally and keeping groupid in server contracts. 
I really don't like the name lgid. It really does not tell me anything and feels confusing. Let's just have groupId on the server contracts.

File: packages/runtime/container-runtime/README.md (hunk 11 above)
Comment (rajatch-ff) [reply packages/runtime/container-runtime/README.md]: discussed on chat.

File: packages/runtime/container-runtime/README.md (hunk 10 above)
Comment (rajatch-ff) [reply packages/runtime/container-runtime/README.md]: This statement is still redundant.

File: packages/runtime/container-runtime/README.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (rajatch-ff) [reply packages/runtime/container-runtime/README.md:80]: discussed on chat. Thanks for updating.

**Review by rajatch5 — APPROVED** [review]

File: packages/runtime/runtime-definitions/src/dataStoreContext.ts (hunk 2 above)
Comment (jatgarg) [reply packages/runtime/runtime-definitions/src/dataStoreContext.ts:199]: Reduced the comments here.

File: packages/runtime/runtime-definitions/src/dataStoreContext.ts
Comment (ChumpChief) [inline packages/runtime/runtime-definitions/src/dataStoreContext.ts:1]: Also need to rename the member on IFluidDataStoreContext (Line 400 below) - please also convert its comment to tsdoc instead of inline comment.

**Review by ChumpChief — COMMENTED** [review]
Please also update the doc comments in protocol-definitions (storage.ts and summary.ts) to indicate:
1. "groupId" is the same as "loadingGroupId" seen elsewhere in the repo
2. The naming discrepancy is intentional to minimize snapshot size

File: packages/runtime/runtime-definitions/src/dataStoreContext.ts [hunk 16]
```diff
@@ -396,8 +396,11 @@ export interface IFluidDataStoreContext
 	readonly logger: ITelemetryBaseLogger;
 	readonly clientDetails: IClientDetails;
 	readonly idCompressor?: IIdCompressor;
-	// Represents the group to which the data store belongs too.
-	readonly groupId?: string;
+	/**
+	 * Represents the group to which the data store belongs too. Please refer to this readme for more context.
```
Comment (ChumpChief) [inline packages/runtime/runtime-definitions/src/dataStoreContext.ts:400]:
```suggestion
	 * Represents the loading group to which the data store belongs to. Please refer to this readme for more context.
```

**Review by ChumpChief — APPROVED** [review]
Thank you!
